mirrord/cli/src/vpn.rs

```diff
@@ -0,0 +1,302 @@
+use std::{
+    sync::atomic::{AtomicUsize, Ordering},
+    time::Duration,
+};
+
+use futures::{SinkExt, StreamExt};
+use humansize::{format_size, DECIMAL};
+use k8s_openapi::api::core::v1::ConfigMap;
+use mirrord_analytics::{AnalyticsError, NullReporter, Reporter};
+use mirrord_config::LayerConfig;
+use mirrord_kube::api::kubernetes::create_kube_config;
+use mirrord_progress::{Progress, ProgressTracker};
+use mirrord_protocol::{
+    vpn::{ClientVpn, ServerVpn},
+    ClientMessage, DaemonMessage,
+};
+use mirrord_vpn::config::VpnConfig;
+use tokio::signal;
+
+use crate::{
+    config::VpnArgs,
+    connection::{create_and_connect, AgentConnection},
+    error::{CliError, Result},
+};
+
+async fn agent_request<T>(
+    connection: &mut AgentConnection,
+    request: ClientMessage,
+    mapper: impl Fn(DaemonMessage) -> Option<T>,
+) -> Option<T> {
+    tracing::debug!(client_message = ?request, "out");
+    connection.sender.send(request).await.ok()?;
+
+    connection.receiver.recv().await.and_then(mapper)
+}
+
+#[cfg(not(target_os = "macos"))]
+async fn agent_fetch_file<const B: u64>(
+    connection: &mut AgentConnection,
+    path: std::path::PathBuf,
+) -> mirrord_protocol::RemoteResult<Vec<u8>> {
+    use mirrord_protocol::{file::*, FileRequest, FileResponse};
+
+    let request = FileRequest::Open(OpenFileRequest {
+        path,
+        open_options: OpenOptionsInternal {
+            read: true,
+            ..Default::default()
+        },
+    });
+
+    let Some(FileResponse::Open(response)) = agent_request(
+        connection,
+        ClientMessage::FileRequest(request),
+        get_file_response,
+    )
+    .await
+    else {
+        todo!()
+    };
+
+    let OpenFileResponse { fd } = response?;
+
+    let request = FileRequest::Read(ReadFileRequest {
+        remote_fd: fd,
+        buffer_size: B,
+    });
+
+    let Some(FileResponse::Read(response)) = agent_request(
+        connection,
+        ClientMessage::FileRequest(request),
+        get_file_response,
+    )
+    .await
+    else {
+        todo!();
+    };
+
+    let ReadFileResponse { bytes, .. } = response?;
+
+    let request = FileRequest::Close(CloseFileRequest { fd });
+
+    let _ = connection
+        .sender
+        .send(ClientMessage::FileRequest(request))
+        .await;
+
+    Ok(bytes)
+}
+
+fn get_server_vpn(message: DaemonMessage) -> Option<ServerVpn> {
+    match message {
+        DaemonMessage::Vpn(response) => Some(response),
+        _ => None,
+    }
+}
+
+#[cfg(not(target_os = "macos"))]
+fn get_file_response(message: DaemonMessage) -> Option<mirrord_protocol::FileResponse> {
+    match message {
+        DaemonMessage::File(response) => Some(response),
+        _ => None,
+    }
+}
+
+#[allow(clippy::indexing_slicing)]
+pub async fn vpn_command(args: VpnArgs) -> Result<()> {
```

Comment:
I think we're missing some initial `mirrord-protocol` version negotiation. We should first check that the agent in use is capable of processing vpn messages